File: retaRust/python_arch_reference/reta_architecture/input_semantics.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Dict, Iterable, List, Optional, Sequence, Set

from .schema import RetaContextSchema


COMMA_SPLIT_PATTERN = r",(?![^\[\]\{\}\(\)]*[\]\}\)])"


@dataclass
class RowRangeSyntax:
    multiple_prefix: str = "v"
    comma_split_pattern: str = COMMA_SPLIT_PATTERN

    @classmethod
    def from_schema(cls, schema: Optional[RetaContextSchema], i18n=None) -> "RowRangeSyntax":
        prefix = "v"
        if i18n is not None:
            try:
                prefix = str(i18n.befehle2["v"])
            except Exception:
                prefix = "v"
        return cls(multiple_prefix=prefix)

    @classmethod
    def from_i18n(cls, i18n) -> "RowRangeSyntax":
        return cls.from_schema(None, i18n=i18n)

    def split_comma_list(self, text: str) -> List[str]:
        return [segment for segment in re.split(self.comma_split_pattern, str(text))]

    def compact_comma_list(self, text: str) -> str:
        return ",".join(segment for segment in self.split_comma_list(text) if segment)

    def integer_range_pattern(self) -> re.Pattern:
        return re.compile(
            rf"^({re.escape(self.multiple_prefix)}?-?\d+)(-\d+)?((\+)(\d+))*$"
        )

    def fraction_range_pattern(self) -> re.Pattern:
        return re.compile(
            rf"^({re.escape(self.multiple_prefix)}?-?\d+/\d+)(-\d+/\d+)?((\+)(\d+/\d+))*$"
        )

    def is_integer_range_token(self, text: str) -> bool:
        return bool(re.fullmatch(self.integer_range_pattern(), str(text)))

    def is_fraction_range_token(self, text: str) -> bool:
        return bool(re.fullmatch(self.fraction_range_pattern(), str(text)))

    def snapshot(self) -> Dict[str, object]:
        return {
            "multiple_prefix": self.multiple_prefix,
            "comma_split_pattern": self.comma_split_pattern,
        }


@dataclass
class PromptVocabulary:
    main_parameters: List[str] = field(default_factory=list)
    spalten: List[str] = field(default_factory=list)
    eigs_n: List[str] = field(default_factory=list)
    eigs_r: List[str] = field(default_factory=list)
    spalten_dict: Dict[str, List[str]] = field(default_factory=dict)
    ausgabe_paras: List[str] = field(default_factory=list)
    kombi_main_paras: List[str] = field(default_factory=list)
    zeilen_paras: List[str] = field(default_factory=list)
    haupt_for_neben: List[str] = field(default_factory=list)
    not_parameter_values: List[str] = field(default_factory=list)
    haupt_for_neben_set: Set[str] = field(default_factory=set)
    ausgabe_art: List[str] = field(default_factory=list)
    zeilen_typen: List[str] = field(default_factory=list)
    zeilen_zeit: List[str] = field(default_factory=list)
    zeilen_typen_b: List[str] = field(default_factory=list)
    gebrochen_erlaubte_zahlen: Set[int] = field(default_factory=set)
    befehle: List[str] = field(default_factory=list)
    befehle2: Set[str] = field(default_factory=set)

    def snapshot(self) -> Dict[str, object]:
        return {
            "main_parameters_len": len(self.main_parameters),
            "spalten_len": len(self.spalten),
            "spalten_dict_keys": len(self.spalten_dict),
            "ausgabe_paras_len": len(self.ausgabe_paras),
            "kombi_main_paras_len": len(self.kombi_main_paras),
            "zeilen_paras_len": len(self.zeilen_paras),
            "haupt_for_neben_len": len(self.haupt_for_neben),
            "ausgabe_art_len": len(self.ausgabe_art),
            "befehle_len": len(self.befehle),
            "befehle2_len": len(self.befehle2),
            "gebrochen_erlaubte_zahlen_len": len(self.gebrochen_erlaubte_zahlen),
        }


class PromptVocabularyBuilder:
    def __init__(self, schema: RetaContextSchema, row_ranges: Optional[RowRangeSyntax] = None) -> None:
        self.schema = schema
        self.row_ranges = row_ranges or RowRangeSyntax.from_schema(schema)

    def build(self, program, i18n) -> PromptVocabulary:
        main_parameters = ["-" + str(a) for a in program.mainParaCmds]
        spalten = ["--" + str(a[0]) + "=" for a in program.paraDict.keys()] + ["--="]

        eigs_n: List[str] = []
        eigs_r: List[str] = []
        for pp in program.paraDict.keys():
            if pp[0] == i18n.konzeptE["konzept"]:
                eigs_n.append(pp[1])
            elif pp[0] == i18n.konzeptE["konzept2"]:
                eigs_r.append(pp[1])

        spalten_dict: Dict[str, List[str]] = {}
        for tupel in program.paraNdataMatrix:
            for haupt in tupel[0]:
                spalten_dict.setdefault(haupt, []).extend(list(tupel[1]))
                spalten_dict.setdefault("*", []).extend(list(tupel[1]))

        spalten.extend(
            [
                "--" + i18n.ausgabeParas["breite"] + "=",
                "--" + i18n.ausgabeParas["breiten"] + "=",
                "--" + i18n.ausgabeParas["keinenummerierung"],
                "--*=",
            ]
        )

        zeilen_typen = [
            i18n.zeilenParas["sonne"],
            i18n.zeilenParas["mond"],
            i18n.zeilenParas["planet"],
            i18n.zeilenParas["schwarzesonne"],
            i18n.zeilenParas["SonneMitMondanteil"],
            "*",
        ]
        zeilen_zeit = [
            i18n.zeilenParas["heute"],
            i18n.zeilenParas["gestern"],
            i18n.zeilenParas["morgen"],
            "*",
        ]
        zeilen_typen_b = [
            i18n.zeilenParas["aussenerste"],
            i18n.zeilenParas["innenerste"],
            i18n.zeilenParas["aussenalle"],
            i18n.zeilenParas["innenalle"],
            "*",
        ]

        ausgabe_paras = [
            "--" + s + ("=" if l else "")
            for s, l in zip(i18n.ausgabeParas.values(), i18n.ausgabeParasEqSign.values())
        ] + ["--*="]

        kombi_main_paras = [
            "--" + i18n.kombiMainParas["galaxie"] + "=",
            "--" + i18n.kombiMainParas["universum"] + "=",
            "--*=",
        ]

        zeilen_paras = [
            "--" + i18n.zeilenParas["zeit"] + "=",
            "--" + i18n.zeilenParas["zaehlung"] + "=",
            "--" + i18n.zeilenParas["vorhervonausschnitt"] + "=",
            "--" + i18n.zeilenParas["vorhervonausschnittteiler"],
            "--" + i18n.zeilenParas["primzahlvielfache"] + "=",
            "--" + i18n.zeilenParas["nachtraeglichneuabzaehlung"] + "=",
            "--" + i18n.zeilenParas["nachtraeglichneuabzaehlungvielfache"] + "=",
            "--" + i18n.zeilenParas["alles"],
            "--" + i18n.zeilenParas["potenzenvonzahlen"] + "=",
            "--" + i18n.zeilenParas["typ"] + "=",
            "--" + i18n.zeilenParas["vielfachevonzahlen"] + "=",
            "--" + i18n.zeilenParas["oberesmaximum"] + "=",
            "--" + i18n.zeilenParas["primzahlen"] + "=",
            "--" + i18n.zeilenParas["invertieren"],
            "--*=",
        ]

        haupt_for_neben = [
            "-" + s
            for s in set(i18n.hauptForNeben.values()) - {i18n.mainParaCmds["debug"]}
        ]
        not_parameter_values = ausgabe_paras + zeilen_paras + kombi_main_paras + spalten + main_parameters
        haupt_for_neben_set = set(haupt_for_neben)
        ausgabe_art = list(i18n.ausgabeArt.values())

        zum_vergleich: List[str] = []
        gebrochen_erlaubte_zahlen: Set[int] = set()
        for a in program.paraNdataMatrix:
            for b in a[1]:
                zum_vergleich.append(b)
                if len(set(a[0]) & i18n.gebrochenUniGalEinzeln) > 0:
                    gebrochen_erlaubte_zahlen |= {int(b)}
        if gebrochen_erlaubte_zahlen:
            gebrochen_erlaubte_zahlen -= {max(gebrochen_erlaubte_zahlen)}

        befehle = list(i18n.befehle)
        befehle += [i18n.EIGS_N_R[0] + a for a in eigs_n] + [i18n.EIGS_N_R[1] + a for a in eigs_r]
        befehle2 = set(befehle) - {"reta"}

        return PromptVocabulary(
            main_parameters=main_parameters,
            spalten=spalten,
            eigs_n=eigs_n,
            eigs_r=eigs_r,
            spalten_dict=spalten_dict,
            ausgabe_paras=ausgabe_paras,
            kombi_main_paras=kombi_main_paras,
            zeilen_paras=zeilen_paras,
            haupt_for_neben=haupt_for_neben,
            not_parameter_values=not_parameter_values,
            haupt_for_neben_set=haupt_for_neben_set,
            ausgabe_art=ausgabe_art,
            zeilen_typen=zeilen_typen,
            zeilen_zeit=zeilen_zeit,
            zeilen_typen_b=zeilen_typen_b,
            gebrochen_erlaubte_zahlen=gebrochen_erlaubte_zahlen,
            befehle=befehle,
            befehle2=befehle2,
        )


@dataclass
class InputBundle:
    schema: RetaContextSchema
    row_ranges: RowRangeSyntax
    prompt_vocabulary_builder: PromptVocabularyBuilder

    @classmethod
    def from_schema(cls, schema: RetaContextSchema, i18n=None) -> "InputBundle":
        row_ranges = RowRangeSyntax.from_schema(schema, i18n=i18n)
        return cls(
            schema=schema,
            row_ranges=row_ranges,
            prompt_vocabulary_builder=PromptVocabularyBuilder(schema, row_ranges=row_ranges),
        )

    def build_prompt_vocabulary(self, program, i18n) -> PromptVocabulary:
        return self.prompt_vocabulary_builder.build(program, i18n)

    def snapshot(self) -> Dict[str, object]:
        return {
            "row_ranges": self.row_ranges.snapshot(),
            "prompt_vocabulary_builder": {
                "available": True,
            },
        }

File: retaRust/crates/reta_architecture/src/input_semantics.rs
//! Input semantics transcompiled from
//! `python_arch_reference/reta_architecture/input_semantics.py`.
//!
//! This bridges schema aliases, row-range syntax and prompt vocabulary in one
//! typed bundle.  Python builds the concrete lists from runtime i18n objects;
//! Rust stores the same vocabulary surface and can be filled from the split
//! schema or by later generated i18n loaders.

use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};

use crate::row_ranges::{bootstrap_row_range_morphisms, RowRangeMorphismBundle, RowRangeSyntax};
use crate::schema::RetaContextSchema;

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PromptVocabulary {
    pub main_parameters: Vec<String>,
    pub spalten: Vec<String>,
    pub eigs_n: Vec<String>,
    pub eigs_r: Vec<String>,
    pub spalten_dict: BTreeMap<String, Vec<String>>,
    pub ausgabe_paras: Vec<String>,
    pub kombi_main_paras: Vec<String>,
    pub zeilen_paras: Vec<String>,
    pub haupt_for_neben: Vec<String>,
    pub not_parameter_values: Vec<String>,
    pub haupt_for_neben_set: BTreeSet<String>,
    pub ausgabe_art: Vec<String>,
    pub zeilen_typen: Vec<String>,
    pub zeilen_zeit: Vec<String>,
    pub zeilen_typen_b: Vec<String>,
    pub gebrochen_erlaubte_zahlen: BTreeSet<i64>,
    pub befehle: Vec<String>,
    pub befehle2: BTreeSet<String>,
}

impl PromptVocabulary {
    pub fn snapshot(&self) -> PromptVocabularySnapshot {
        PromptVocabularySnapshot {
            main_parameters_len: self.main_parameters.len(),
            spalten_len: self.spalten.len(),
            spalten_dict_keys: self.spalten_dict.len(),
            ausgabe_paras_len: self.ausgabe_paras.len(),
            kombi_main_paras_len: self.kombi_main_paras.len(),
            zeilen_paras_len: self.zeilen_paras.len(),
            haupt_for_neben_len: self.haupt_for_neben.len(),
            ausgabe_art_len: self.ausgabe_art.len(),
            befehle_len: self.befehle.len(),
            befehle2_len: self.befehle2.len(),
            gebrochen_erlaubte_zahlen_len: self.gebrochen_erlaubte_zahlen.len(),
        }
    }
}

impl Default for PromptVocabulary {
    fn default() -> Self {
        Self {
            main_parameters: Vec::new(),
            spalten: Vec::new(),
            eigs_n: Vec::new(),
            eigs_r: Vec::new(),
            spalten_dict: BTreeMap::new(),
            ausgabe_paras: Vec::new(),
            kombi_main_paras: Vec::new(),
            zeilen_paras: Vec::new(),
            haupt_for_neben: Vec::new(),
            not_parameter_values: Vec::new(),
            haupt_for_neben_set: BTreeSet::new(),
            ausgabe_art: Vec::new(),
            zeilen_typen: Vec::new(),
            zeilen_zeit: Vec::new(),
            zeilen_typen_b: Vec::new(),
            gebrochen_erlaubte_zahlen: BTreeSet::new(),
            befehle: Vec::new(),
            befehle2: BTreeSet::new(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PromptVocabularySnapshot {
    pub main_parameters_len: usize,
    pub spalten_len: usize,
    pub spalten_dict_keys: usize,
    pub ausgabe_paras_len: usize,
    pub kombi_main_paras_len: usize,
    pub zeilen_paras_len: usize,
    pub haupt_for_neben_len: usize,
    pub ausgabe_art_len: usize,
    pub befehle_len: usize,
    pub befehle2_len: usize,
    pub gebrochen_erlaubte_zahlen_len: usize,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PromptVocabularyBuilder {
    pub schema: RetaContextSchema,
    pub row_ranges: RowRangeSyntax,
}

impl PromptVocabularyBuilder {
    pub fn new(schema: RetaContextSchema, row_ranges: RowRangeSyntax) -> Self {
        Self { schema, row_ranges }
    }

    pub fn build_from_schema(&self) -> PromptVocabulary {
        let main_parameters = self
            .schema
            .main_alias_groups()
            .into_iter()
            .map(|group| format!("-{}", group.canonical))
            .collect::<Vec<_>>();

        let mut spalten = self
            .schema
            .sub_parameter_alias_groups()
            .keys()
            .map(|main| format!("--{main}="))
            .collect::<Vec<_>>();
        if !spalten.iter().any(|value| value == "--=") {
            spalten.push("--=".to_string());
        }

        let ausgabe_paras = self
            .schema
            .output_parameters
            .values()
            .map(|value| format!("--{value}="))
            .chain(std::iter::once("--*=".to_string()))
            .collect::<Vec<_>>();
        let kombi_main_paras = self
            .schema
            .combination_parameters
            .values()
            .map(|value| format!("--{value}="))
            .chain(std::iter::once("--*=".to_string()))
            .collect::<Vec<_>>();
        let zeilen_paras = self
            .schema
            .row_parameters
            .values()
            .map(|value| format!("--{value}="))
            .chain(std::iter::once("--*=".to_string()))
            .collect::<Vec<_>>();
        let haupt_for_neben = self
            .schema
            .scopes
            .values()
            .map(|value| format!("-{value}"))
            .collect::<Vec<_>>();
        let haupt_for_neben_set = haupt_for_neben.iter().cloned().collect::<BTreeSet<_>>();
        let ausgabe_art = self
            .schema
            .output_modes
            .values()
            .cloned()
            .collect::<Vec<_>>();

        let mut not_parameter_values = Vec::new();
        not_parameter_values.extend(ausgabe_paras.iter().cloned());
        not_parameter_values.extend(zeilen_paras.iter().cloned());
        not_parameter_values.extend(kombi_main_paras.iter().cloned());
        not_parameter_values.extend(spalten.iter().cloned());
        not_parameter_values.extend(main_parameters.iter().cloned());

        let mut befehle = vec![
            "reta".to_string(),
            "exit".to_string(),
            "quit".to_string(),
            "help".to_string(),
        ];
        befehle.extend(main_parameters.iter().cloned());
        let befehle2 = befehle
            .iter()
            .filter(|cmd| cmd.as_str() != "reta")
            .cloned()
            .collect();

        PromptVocabulary {
            main_parameters,
            spalten,
            ausgabe_paras,
            kombi_main_paras,
            zeilen_paras,
            haupt_for_neben,
            not_parameter_values,
            haupt_for_neben_set,
            ausgabe_art,
            zeilen_typen: vec![
                "sonne".to_string(),
                "mond".to_string(),
                "planet".to_string(),
                "*".to_string(),
            ],
            zeilen_zeit: vec![
                "heute".to_string(),
                "gestern".to_string(),
                "morgen".to_string(),
                "*".to_string(),
            ],
            zeilen_typen_b: vec![
                "aussenerste".to_string(),
                "innenerste".to_string(),
                "aussenalle".to_string(),
                "innenalle".to_string(),
                "*".to_string(),
            ],
            befehle,
            befehle2,
            ..PromptVocabulary::default()
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct InputBundle {
    pub schema: RetaContextSchema,
    pub row_ranges: RowRangeMorphismBundle,
    pub prompt_vocabulary_builder: PromptVocabularyBuilder,
}

impl InputBundle {
    pub fn from_schema(schema: RetaContextSchema, syntax: Option<RowRangeSyntax>) -> Self {
        let row_ranges = bootstrap_row_range_morphisms(syntax);
        let prompt_vocabulary_builder =
            PromptVocabularyBuilder::new(schema.clone(), row_ranges.syntax.clone());
        Self {
            schema,
            row_ranges,
            prompt_vocabulary_builder,
        }
    }

    pub fn build_prompt_vocabulary(&self) -> PromptVocabulary {
        self.prompt_vocabulary_builder.build_from_schema()
    }

    pub fn snapshot(&self) -> InputBundleSnapshot {
        InputBundleSnapshot {
            row_ranges_multiple_prefix: self.row_ranges.syntax.multiple_prefix.clone(),
            row_ranges_stage: self.row_ranges.activated_stage,
            prompt_vocabulary_builder_available: true,
            schema_main_alias_groups: self.schema.main_alias_groups().len(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct InputBundleSnapshot {
    pub row_ranges_multiple_prefix: String,
    pub row_ranges_stage: u32,
    pub prompt_vocabulary_builder_available: bool,
    pub schema_main_alias_groups: usize,
}

pub fn bootstrap_input_semantics(schema: Option<RetaContextSchema>) -> InputBundle {
    InputBundle::from_schema(schema.unwrap_or_else(crate::schema::bootstrap_schema), None)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prompt_vocabulary_counts_main_aliases() {
        let bundle = bootstrap_input_semantics(None);
        let vocab = bundle.build_prompt_vocabulary();
        assert!(vocab
            .main_parameters
            .iter()
            .any(|value| value == "-spalten"));
        assert!(bundle.snapshot().prompt_vocabulary_builder_available);
    }
}
